GTG/core/datastore2.py

```diff
@@ -0,0 +1,384 @@
+# this program.  If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+"""The datastore ties together all the basic type stores and backends."""
+
+import os
+import threading
+import logging
+import shutil
+from datetime import datetime
+from time import time
+
+from GTG.core.tasks2 import TaskStore
+from GTG.core.tags2 import TagStore
+from GTG.core.saved_searches import SavedSearchStore
+from GTG.core import firstrun_tasks
+from GTG.backends.backend_signals import BackendSignals
+
+from lxml import etree as et
+
+
+log = logging.getLogger(__name__)
+
+
+class Datastore2:
+
+    #: Amount of backups to keep
+    BACKUPS_NUMBER = 7
+
+
+    def __init__(self) -> None:
+        self.tasks = TaskStore()
+        self.tags = TagStore()
+        self.saved_searches = SavedSearchStore()
+        self.xml_tree = None
+
+        self._mutex = threading.Lock()
+        self.backends = {}
+        self._backend_signals = None
+
+        # Flag when turned to true, all pending operation should be
+        # completed and then GTG should quit
+        self.please_quit = False
+
+
+    @property
+    def mutex(self) -> threading.Lock:
+        return self._mutex
+
+
+    def load_data(self, data: et.Element) -> None:
+
+        self.saved_searches.from_xml(data.find('searchlist'))
+        self.tags.from_xml(data.find('taglist'))
+        self.tasks.from_xml(data.find('tasklist'), self.tags)
+
+
+    def load_file(self, path: str) -> None:
+
+        if log.isEnabledFor(logging.DEBUG):
+            bench_start = time()
+
+        parser = et.XMLParser(remove_blank_text=True, strip_cdata=False)
+
+        with open(path, 'rb') as stream:
+            self.tree = et.parse(stream, parser=parser)
+            self.load_data(self.tree)
+
+        if log.isEnabledFor(logging.DEBUG):
+            log.debug('Processed file %s in %.2fms',
+                      path, (time() - bench_start) * 1000)
+
+
+    def generate_xml(self) -> et.ElementTree:
+
+        root = et.Element('gtgData')
+        root.set('appVersion', '0.5')
+        root.set('xmlVersion', '2')
+
+        root.append(self.tags.to_xml())
+        root.append(self.saved_searches.to_xml())
+        root.append(self.tasks.to_xml())
+
+        return et.ElementTree(root)
+
+
+    def save_file(self, path: str) -> None:
+
+        temp_file = path + '__'
+
+        if os.path.exists(path):
+            os.rename(path, temp_file)
+
+        if log.isEnabledFor(logging.DEBUG):
+            bench_start = time()
+
+        tree = self.generate_xml()
+
+        try:
+            with open(path, 'wb') as stream:
+                tree.write(stream, xml_declaration=True,
+                        pretty_print=True,
+                        encoding='UTF-8')
+
+        except (IOError, FileNotFoundError):
+            log.error('Could not write XML file at %r', path)
```

Comment:
I am wondering about passing errors. Like imagine if you exit GTG and for some reason you can't write the data. We could then display a dialog box telling the user that and the user could decide not to exit GTG to maybe export the precious tasks somewhere else.